invokeai/app/invocations/facetools.py

```diff
@@ -0,0 +1,695 @@
+import math
+from typing import TypedDict
+
+import cv2
+import mediapipe as mp
+import numpy as np
+from PIL import Image, ImageDraw, ImageFilter, ImageOps
+from PIL.Image import Image as ImageType
+
+from invokeai.app.invocations.baseinvocation import (
+    BaseInvocation,
+    InputField,
+    InvocationContext,
+    OutputField,
+    invocation,
+    invocation_output,
+)
+from invokeai.app.invocations.primitives import ImageField, ImageOutput
+from invokeai.app.models.image import ImageCategory, ResourceOrigin
+
+
+@invocation_output("face_mask_output")
+class FaceMaskOutput(ImageOutput):
+    """Base class for FaceMask output"""
+
+    mask: ImageField = OutputField(description="The output mask")
+
+
+@invocation_output("face_off_output")
+class FaceOffOutput(ImageOutput):
+    """Base class for FaceOff Output"""
+
+    mask: ImageField = OutputField(description="The output mask")
+    x: int = OutputField(description="The x coordinate of the bounding box's left side")
+    y: int = OutputField(description="The y coordinate of the bounding box's top side")
+
+
+class FaceResultsData(TypedDict):
+    pil_image: ImageType
+    mask_pil: ImageType
```

Comment:
Based on the usage of `FaceResultsData` below, I think `pil_image` and `mask_pil` may actually be optional, in which case they should be typed here as `Optional[ImageType]`. Not totally clear to me tho